pkg/hcl_interpreter/term.go

```diff
@@ -334,8 +327,8 @@ func (t Term) Evaluate(
 					forEachResult.add(k, evalWithEach(k, v))
 				}
 			}
-			return forEachResult.value(), diagnostics
 		}
+		return forEachResult.value(), diagnostics
```

Comment:
If I understand the PR description correctly, we return an "empty object" for `(&forEachResult{}).value()`.  Could you explain how that fixes the bug?

Reply:
I think it's easiest to understand this is we take the perspective of the caller of this method (the `hcl_interpreter.go` file).

The caller of this method expects to receive some value representing one or more resources; depending on if `for_each` / `count` is set.  This expectation is expressed by the `Multiple` field in `ResourceMeta`.

In the case of this bug, we were returning a single resource, e.g.:

```json
{"bucket_prefix": "foo", "tags": {"Name": "My bucket"}}
```

However, because `Multiple` is set, the caller really expects something of this shape:

1.  An array for `count`:

    ```json
    [{"bucket_prefix": "foo", "tags": {"Name": "My bucket"}}]
    ```
    
2.  An object for `for_each`:

    ```json
    {"my_bucket": {"bucket_prefix": "foo", "tags": {"Name": "My bucket"}}}
    ```

So, given our expectation for something like (2), we are interpreting the single resource as an object-of-resources; with two resources, `bucket_prefix` and `tags`.  This leads to the bizarre output.

There are really two options to fix this:

1.  When things go wrong for `count`, we set it to 1.  We could do something similar here and invent a key, then return that:

    ```json
    {"invented_key": {"bucket_prefix": "foo", "tags": {"Name": "My bucket"}}}
    ```
    
    But what should `invented_key` be?  It's a lot more dubious than count, since there it should obviously be `0`.
    
2.  We can choose to return no objects, so the object-of-resources is empty:

    ```json
    {}
    ```
    
This choice is a bit subjective and I went with (2).  Keep in mind that this is static analysis where some part of the code already failed to evaluate, so there's no real good answer and we just have to figure out something sensible to approach what the user intended...

Reply:
Oh okay, now I get it! 

> But what should invented_key be? It's a lot more dubious than count, since there it should obviously be 0.

Mh, couldn't we just introduce some kind of placeholder? 

I guess the main question is: does it matter? Do we produce a better analysis if we have that resource? 

And if I understand your last sentence correctly, it doesn't :)

Reply:
> Mh, couldn't we just introduce some kind of placeholder?
>
> Do we produce a better analysis if we have that resource?

Yes, I think we could introduce a placeholder.  Our analysis wouldn't necessarily be better, but if we have rules for that resource type, they may detect invalid settings (this is sort of why we're setting `count = 1`).  Feel free to open up a separate ticket for this.